src/lib/mdns/minimal/responders/QueryResponder.h

```diff
@@ -0,0 +1,225 @@
+{
+    bool reportNowAsAdditional; // report as additional data required
+
+    bool alsoReportAdditionalQName = false; // report more data when this record is listed
+    FullQName additionalQName;
+
+    void Clear()
+    {
+        responder                 = nullptr;
+        reportService             = false;
+        reportNowAsAdditional     = false;
+        alsoReportAdditionalQName = false;
+    }
+};
+
+} // namespace Internal
+
+/// Allows building query responder configuration
+class QueryResponderSettings
+{
+public:
+    QueryResponderSettings() : mInfo(nullptr) {}
+    QueryResponderSettings(Internal::QueryResponderInfo * info) : mInfo(info) {}
+    QueryResponderSettings(const QueryResponderSettings & other) = default;
+
+    QueryResponderSettings & SetReportInServiceListing(bool reportService)
+    {
+        if (IsValid())
+        {
+            mInfo->reportService = reportService;
+        }
+        return *this;
+    }
+
+    QueryResponderSettings & SetReportAdditional(const FullQName & qname)
+    {
+        if (IsValid())
+        {
+            mInfo->alsoReportAdditionalQName = true;
+            mInfo->additionalQName           = qname;
+        }
+        return *this;
+    }
+
+    bool IsValid() const { return mInfo != nullptr; }
+
+private:
+    Internal::QueryResponderInfo * mInfo;
+};
+
+class QueryResponderIterator : public std::iterator<std::input_iterator_tag, QueryResponderRecord>
+{
+public:
+    QueryResponderIterator() : mAdditionalOnly(false), mCurrent(nullptr), mRemaining(0) {}
+    QueryResponderIterator(bool additionalOnly, Internal::QueryResponderInfo * pos, size_t size) :
+        mAdditionalOnly(additionalOnly), mCurrent(pos), mRemaining(size)
+    {
+        SkipInvalid();
+    }
+    QueryResponderIterator(const QueryResponderIterator & other) = default;
+    QueryResponderIterator & operator=(const QueryResponderIterator & other) = default;
+
+    QueryResponderIterator & operator++()
+    {
+        if (mRemaining != 0)
+        {
+            mCurrent++;
+            mRemaining--;
+        }
+        SkipInvalid();
+        return *this;
+    }
+
+    QueryResponderIterator operator++(int)
+    {
+        QueryResponderIterator tmp(*this);
+        operator++();
+        return tmp;
+    }
+
+    bool operator==(const QueryResponderIterator & rhs) const { return mCurrent == rhs.mCurrent; }
+    bool operator!=(const QueryResponderIterator & rhs) const { return mCurrent != rhs.mCurrent; }
+
+    QueryResponderRecord & operator*() { return *mCurrent; }
+    QueryResponderRecord * operator->() { return mCurrent; }
+
+    Internal::QueryResponderInfo * GetInternal() { return mCurrent; }
+    const Internal::QueryResponderInfo * GetInternal() const { return mCurrent; }
+
+private:
+    bool isCurrentValid() const
+    {
+        if (mCurrent->responder == nullptr)
+        {
+            return false;
+        }
+
+        if (mAdditionalOnly && !mCurrent->reportNowAsAdditional)
```

Comment:
It's not super clear by the time we read the code here what `mCurrent->reportNowAsAdditional` means and why it would invalidate `isCurrentValid()`

Reply:
Added a comment for the method itself. Not sure if that useful since it kind of restates the code .